### DISCUSSION TITLE
Setting Address for Multivendor Order

### DISCUSSION THREAD

**Discussion by dylviz** [vendurehq/vendure#2167] [Q&A] [answered]
I'm playing with the Multivendor plugin but I have a question that I don't think is a bug or related to the multivendor plugin. I'm having some trouble getting to the payment process after pulling together a cart from two different sellers.

Here is some info about what I did:

### Create a new customer:
```
mutation {
  registerCustomerAccount(
    input: {
      emailAddress: "nnn@nnn.com",
      password: "password",
      firstName: "John",
      lastName: "Doe"
    }
  ) {
   ... on Success {
        success
    }
    ... on ErrorResult {
        errorCode
        message
    }
    ... on PasswordValidationError {
        validationErrorMessage
    }
  }
}
```

### Login with the new customer:
```
mutation {
    login(username: "nnn@nnn.com", password: "password", rememberMe: true){
        ... on CurrentUser{
            id
            identifier
            channels{
                id
                token
                permissions
            }
        }
        ... on ErrorResult {
            errorCode
            message
        }
    }
}
```

### Create an Order
```
mutation MultivendorCheckout{
    authenticate(
        input: { native: { username: "nnn@nnn.com", password: "password"}}
    ) {
        ... on CurrentUser{
            id
        }
    }
    addItemToOrder(productVariantId: "92", quantity: 1,){
        __typename
        ... on Order{
            id
        }
        ... on ErrorResult{
            errorCode
            errorCode
        }
    }
    add2: addItemToOrder(productVariantId: "89", quantity: 1, ){
        __typename
        ... on Order{
            id
        }
        ... on ErrorResult{
            errorCode
            errorCode
        }
    }
}
```

### Set Order Shipping Address
```
mutation SetOrderShippingAddress{
    setOrderShippingAddress(
        input: {
            company: null
            streetLine1: "123 Jalan Pasir Puteh"
            streetLine2: "Oakland"
            city: "Epoh"
            province: "Perok"
            postalCode: "123456"
            countryCode: "MV"
            customFields: {
                houseNumber: "42"
            }
        }
    ){
        ... on Order{
            id
        }
    }
}
```

###  Get Shipping methods
```
query availableShippingMethods{
    eligibleShippingMethods{
        id
        name
        priceWithTax
    }
}
```

### Set Shipping Method
```
mutation AssignShipping{
    setOrderShippingMethod(shippingMethodId: ["3", "4"]){
        ... on Order{
            id
        }
    }
}
```

### And Finally Checkout
```
mutation CheckOut{
    transitionOrderToState(state: "ArrangingPayment"){
        ... on Order{
            id
        }
        ... on OrderStateTransitionError{
            errorCode
            message
            transitionError
        }
    }
    payment1: addPaymentToOrder(
        input: { method: "connected-payment-method", metadata: {}}
    ){
        ... on Order{id}
        ... on ErrorResult{
            errorCode
            message
        }
        ... on PaymentFailedError{
            paymentErrorMessage
        }
    }
}
```

### Error:
```
{
    "data": {
        "transitionOrderToState": {
            "errorCode": "ORDER_STATE_TRANSITION_ERROR",
            "message": "Cannot transition Order from \"AddingItems\" to \"ArrangingPayment\"",
            "transitionError": "not all lines have shipping"
        },
        "payment1": {
            "errorCode": "ORDER_PAYMENT_STATE_ERROR",
            "message": "A Payment may only be added when Order is in \"ArrangingPayment\" state"
        }
    }
}
```

I think I'm not setting something correctly but can't find the right way to do it. Any help would be appreciated. Thanks! In the browser, I can see everything looks good. I can see the order and both shipping methods that are set.

**Comment by dylviz** [comment]
After doing some digging, the error looks to be generated in mv-order-process.ts at the lines here:
```
if (fromState === 'AddingItems' && toState === 'ArrangingPayment') {
    for (const line of data.order.lines) {
        if (!line.shippingLineId) {
            return 'not all lines have shipping';
        }
    }
}
```

Still digging...


This is the same problem that appears here during the demo: https://youtu.be/jTOhzrA1baQ?t=901

**Comment by dylviz** [comment]
I just tried starting from the top with a fresh DB, just to double check everything and the problem persists.. This is a tough one..

I notice the ShippingLineID of order_line is null in the DB but just not sure why:
![image](https://user-images.githubusercontent.com/25091990/236974555-96b30e14-576d-4372-9ca5-68a9ced29248.png)

**Comment by tlodge** [comment]
Hmm.  I just tried the same and got the same result - though only created a single order to see if that made any difference.  My feeling is that the problem is that `mv-order-process.ts ` is searching through` order.lines` when I think it should be looking at `order.shippingLines`. When you trace the` setOrderShippingMethod` it calls `async setShippingMethod(ctx, orderId, shippingMethodIds) {`  in the file `vendure/core/dist/service/services/order.service.js` which looks like it updates `order.shippingLines `and does nothing with `order.lines` and so doesn't set `shippingLineId`. Incidentally I get the same issue when I use the remix frontend.  My best guess for now is that the `mv-order-process.ts `is out of date.

**Reply by tlodge** [reply]
So if you change the offending lines in the mv-order-process.ts to:

```
            for (const line of data.order.shippingLines) {
                if (!line.shippingMethodId) {
                    return 'not all lines have shipping';
                }
            }
```

then all works fine.  Not 100% sure this is the correct fix mind you.

**Reply by dylviz** [reply]
Hmm.. interesting fix. So you just removed the outer condition? Was there other adjustments as well? I'll try this out later today

**Reply by tlodge** [reply, marked as answer]
I kept the outer condition but changed the inner to look at shippingLines rather than lines - full method:

```
 async onTransitionStart(fromState, toState, data) {
        const { ctx, order } = data;
        if (fromState === 'AddingItems' && toState === 'ArrangingPayment') {
            for (const line of data.order.shippingLines) {
                if (!line.shippingMethodId) {
                    return 'not all lines have shipping';
                }
            }
        }
```

I didn't change anything else.  I have convinced myself that this could be the right fix, but not 100% sure it won't break something else.  Found another anomaly that I'll add as a new discussion so I don't hijack this thread.

**Reply by tlodge** [reply]
If interested the other issue I'm seeing is logged [here](https://github.com/vendure-ecommerce/vendure/issues/2176)

**Comment by dylviz** [comment]
@tlodge  Sorry was a busy weekend. After doing some tests on my end, I would agree with your fix. I will move to the issue you linked and see if I run into that as well
